### PR TITLE
fix(cgan): make density output activation configurable

### LINKED ISSUE #69: Align cGAN output activation with EngiBench density range
## Problem

The current `cgan_cnn_2d` generator ends with `nn.Tanh()`, so raw generated values are in `[-1, 1]`. However, EngiBench density fields are represented in `[0, 1]`, and validation/evaluation currently clip raw cGAN outputs directly to `[1e-3, 1]` without rescaling.

This means negative generator outputs are collapsed to the density floor rather than being mapped smoothly into the physical density range. In existing artifacts, a nontrivial fraction of generated pixels can be negative before clipping, especially on `beams2d`, so this can affect visual quality and distributional/feasibility metrics.

## Why this matters

Training and evaluation should use the same output convention. Currently the discriminator sees real designs in `[0, 1]`, while the generator architecture is conventionally configured for `[-1, 1]` outputs. Direct clipping is a pragmatic safety step, but it is not a clean output-space convention.

## Suggested fix

For EngiBench density-field tasks, make the cGAN output `[0, 1]` natively:

- Replace the generator final `nn.Tanh()` with `nn.Sigmoid()`.
- Keep evaluation clipping to `[1e-3, 1]` as a safety/physics-floor operation.
- Retrain cGAN checkpoints; do not reinterpret existing `tanh` checkpoints by applying `(x + 1) / 2` only at evaluation time.

## Alternative valid convention

Keep `nn.Tanh()`, but then consistently train and evaluate in `[-1, 1]`:

- Rescale real training designs with `x_scaled = 2*x - 1`.
- Train the discriminator on scaled real designs.
- During validation/evaluation, map generated outputs back via `x_phys = (x_gen + 1) / 2` before clipping/metrics/optimizer calls.

This is standard for image GANs, but it is more invasive than using `Sigmoid()` for density fields.

## Backward compatibility

To avoid silently changing old behavior, consider adding an explicit option such as:

```bash
--generator-output-activation tanh|sigmoid
```

with the current `tanh` behavior preserved as default initially, and `sigmoid` recommended for new EngiBench density-field runs.

### PR DESCRIPTION
**Pull request by SoheylM**
## Summary
- add a backward-compatible tanh|sigmoid generator output activation option for cgan_2d and cgan_cnn_2d
- keep tanh as the default so existing W&B checkpoints load with their original convention
- let new density-field runs opt into native [0, 1] sigmoid outputs and document the flag

Fixes #69

## Tests
- conda run -n engibench pytest -q
- conda run -n engibench ruff check engiopt/cgan_2d/cgan_2d.py engiopt/cgan_2d/evaluate_cgan_2d.py engiopt/cgan_cnn_2d/cgan_cnn_2d.py engiopt/cgan_cnn_2d/evaluate_cgan_cnn_2d.py tests/test_cgan_output_activation.py
- conda run -n engibench ruff format --check engiopt/cgan_2d/cgan_2d.py engiopt/cgan_2d/evaluate_cgan_2d.py engiopt/cgan_cnn_2d/cgan_cnn_2d.py engiopt/cgan_cnn_2d/evaluate_cgan_cnn_2d.py tests/test_cgan_output_activation.py

### CODE REVIEW

**Comment by SoheylM** [comment]
Adding the qualitative Beams2D sanity-check figure for this cGAN output-activation change.

cGAN 2D 50-epoch Beams2D reference comparison

This is a short local smoke test, not a final benchmark: `cgan_2d` was trained for 50 epochs on 2048 Beams2D training samples and compared against optimized test references under the same conditions. The main motivation for this PR shows up clearly: the legacy `tanh` head still emits invalid negative raw densities (`raw_neg=27.7%` here), while the opt-in `sigmoid` head emits native `[0, 1]` densities (`raw_neg=0.0%`) and modestly improves this quick reference MAE (`0.191` vs `0.210`).

File: tests/test_cgan_output_activation.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
+"""Focused checks for configurable cGAN generator output activations."""
+
+from __future__ import annotations
```
Comment (g-braeunlich) [inline tests/test_cgan_output_activation.py]: No longer needed. Also dangerous if you want to use introspection

File: engiopt/cgan_2d/cgan_2d.py [hunk 2]
```diff
@@ -105,6 +116,14 @@ def forward(self, z: th.Tensor, conds: th.Tensor) -> th.Tensor:
         return design.view(design.size(0), *self.design_shape)
 
 
+def _make_output_activation(activation: GeneratorOutputActivation) -> nn.Module:
```
Comment (g-braeunlich) [inline engiopt/cgan_2d/cgan_2d.py:119]: That one seems to be duplicated.
How about moving that to a shared module?

Other suggestion: Use a dict here, e.g.
```py
activations = {"tahnh": nn.Tanh, "sigmoid": nn.Sigmoid}
try:
    return activations[activation]()
except KeyError:
    raise ValueError(f"Unsupported generator output activation: {activation}") from None

File: engiopt/cgan_cnn_2d/cgan_cnn_2d.py [hunk 3]
```diff
@@ -82,13 +87,14 @@ class Generator(nn.Module):
         out_channels (int): Number of output channels in the final image (e.g. 3 for RGB).
     """
 
-    def __init__(
```
Comment (g-braeunlich) [inline engiopt/cgan_cnn_2d/cgan_cnn_2d.py:85]: I would just put that to the ignores in `pyproject.toml`.
Always having to `noqa` that rule is a PITA. Either stick to the rule or ignore it.
In scientific projects that rule does not make much sense, I would say.

**Review by g-braeunlich — CHANGES_REQUESTED** [review]

File: engiopt/cgan_cnn_2d/cgan_cnn_2d.py (hunk 3 above)
Comment (SoheylM) [reply engiopt/cgan_cnn_2d/cgan_cnn_2d.py:85]: Agreed. I added PLR0913 to the project Ruff ignore list in 06fda8e and removed the inline noqa. In these scientific/ML scripts, long constructors and config-heavy APIs are common enough that a project-level ignore is cleaner than local suppressions.

File: tests/test_cgan_output_activation.py (hunk 1 above)
Comment (SoheylM) [reply tests/test_cgan_output_activation.py]: Removed in 06fda8e, thanks. You are right that postponed annotations are not needed in this test, and keeping runtime annotations normal is better for introspection.

File: engiopt/cgan_2d/cgan_2d.py (hunk 2 above)
Comment (SoheylM) [reply engiopt/cgan_2d/cgan_2d.py:119]: I switched this to the dict-based implementation in 06fda8e. I kept the helper local in both cGAN scripts rather than moving it to a shared module because these baselines follow the CleanRL-style self-contained script pattern. For this tiny architecture-local mapping, the small duplication seems preferable to adding another cross-file dependency for readers who want to inspect or copy one baseline script.

**Review by g-braeunlich — APPROVED** [review]
